pepr/privacy/mia.py

```diff
@@ -0,0 +1,1182 @@
+
+            # Classify
+            eval_indices_target_classified.append([])
+            true_labels_target_classified.append([])
+            for j in range(number_classes):
+                # indices for evaluation_i and attack_true_labels
+                classified_i = np.where(attack_test_labels[evaluation_i] == j)
+
+                # Write evaluation indices
+                eval_c = evaluation_i[classified_i]
+                eval_indices_target_classified[i].append(list(eval_c))
+
+                # Write true labels
+                true_labels_classified = attack_true_labels[classified_i]
+                true_labels_target_classified[i].append(list(true_labels_classified))
+
+        target_model_indices = {
+            "indices_per_target": eval_indices_target_classified,
+            "true_attack_labels_per_target": true_labels_target_classified,
+        }
+
+        return target_model_indices
+
+    @staticmethod
+    def _get_target_predictions(target_models, target_data, attack_test_data):
+        """
+        Get classified target model predictions for every target model.
+
+        Parameters
+        ----------
+        target_models : iterable
+            List of target models to attack.
+        target_data : dict
+            Dictionary storing the classified indices and labels of target training and
+            evaluation records.
+        attack_test_data : numpy.ndarray
+            Array of evaluation data.
+
+        Returns
+        -------
+        list
+            List of shape (target_models, classes) containing the classified
+            predictions per target model and data class.
+        """
+        pred_list = []
+        for i, target_model in enumerate(target_models):
+            pred_list.append([])
+            for j in range(len(target_data["indices_per_target"][i])):
+                class_indices = target_data["indices_per_target"][i][j]
+                pred_list[i].append(
+                    target_model.predict(attack_test_data[class_indices])
+                )
+
+        return pred_list
+
+    @staticmethod
+    def _attack_model_evaluation(
+        attack_models,
+        target_predictions,
+        target_data,
+    ):
+        """
+        Evaluate attack models.
+
+        Parameters
+        ----------
+        attack_models : iterable
+            List of trained attack models to evaluate.
+        target_predictions : list
+            List of prediction vectors per target model and attack model.
+        target_data : dict
+            Dictionary storing the classified indices and labels of target training and
+            evaluation records.
+
+        Returns
+        -------
+        dict
+            Dictionary storing the attack model results. A list "per attack model and
+            target model" has the shape (attack model, target model) -> First index
+            specifies the attack model, the second index the target model.
```

Comment:
Really nice explanation how the indices work! I think we should add it to the description of the `attack_results` dict in the attribute section of MIA, because this would also be valuable in the API documentation (the API documentation only shows docstrings of public functions, meaning without the leading underscore).

Reply:
Yep, good idea! Added in 03c76ef